karabo_data/reader.py

```diff
@@ -258,6 +258,34 @@ def add_key(key, value):
         self._keys_cache[source] = res
         return res
 
+
+class DataChunk:
+    """Reference to a contiguous chunk of data for one or more trains."""
+    def __init__(self, file: FileAccess, source, key, first, train_ids, counts):
+        self.file = file
+        self.source = source
+        self.key = key
+        self.first = first
+        self.train_ids = train_ids
+        self.counts = counts
+
+    @property
+    def slice(self):
+        return slice(self.first, self.first + np.sum(self.counts))
+
+    @property
+    def dataset(self):
+        if self.source in self.file.instrument_sources:
+            group = 'INSTRUMENT'
```

Comment:
I know this is pretty unrelated. But how are INSTRUMENT and DETECTOR sources defined? DETECTOR is fast data (pulse resolved) that is 2D per pusle (hence image)? and INSTRUMENT all other fast data? I know that at the moment AGIPD and LPD are the 'only' detectors. But since over the last few month other detectors  have been integrated we might want to add detectors, or at least revisit the definition of what a detector is. I'll try to collect a list of device names of any devices that should fall under the category "DETECTOR" but are INSTRUMENT at the moment, like the Gotthard.

Reply:
There is no DETECTOR group in the files, as far as I know. AGIPD and LPD data are recorded under INSTRUMENT. If this is something you know of that's been changed recently, please point me to an example, because it will probably need prompt work on `karabo_data`.

The top-level groups I'm aware of are INSTRUMENT, CONTROL, RUN, INDEX and METADATA. So far we've ignored RUN because it was duplicating information already in CONTROL (this is set to change as people can be more selective about what is stored in CONTROL). INSTRUMENT and CONTROL are the groups we use for data, and INDEX and METADATA tell us about how to access the data.

Reply:
Sorry for the confusion, I was referring to the detector_sources property in the reader module, not the HDF5 files. At the moment we have 'only' AGIPD and LPD. I got just curious how you would define a detector? Because as yous said the files do not have a detector group. And based on the AGIPD and LPD example we might want to add more 'detectors' to the detector_sources in karabo_data.

Reply:
With hindsight I probably shouldn't have casually used the term 'detectors' like this. This has come up in #95 as well. Detectors can mean a lot of different things.

My mental model at the time was roughly that each experiment would be centred around one big detector, like AGIPD or LPD, which produces the biggest and most essential data. This is about right for SPB, and I think for FXE as well, but some of the other instruments are doing different kinds of experiments which don't seem to be centred around one huge, specialised detector.

It might be easiest to ultimately drop some 'detector' APIs and move towards more specific functionality, rather than trying to generalise over a set of very different detectors.